File: analysis/git-importer/src/main/kotlin/de/richargh/teamcharta/importer/gitmining/app/internal/WorkItemMiner.kt
package de.richargh.teamcharta.importer.gitmining.app.internal

import de.richargh.teamcharta.importer.git.app.api.Author
import de.richargh.teamcharta.importer.git.app.api.Commit
import de.richargh.teamcharta.importer.git.app.api.CommitType
import de.richargh.teamcharta.importer.git.app.api.WorkKey
import de.richargh.teamcharta.importer.gitmining.app.api.AuthorContribution
import de.richargh.teamcharta.importer.gitmining.app.api.WorkItem
import de.richargh.teamcharta.importer.gitmining.app.api.WorkItems
import java.nio.file.Path
import java.time.ZonedDateTime

fun extractWorkItems(commits: List<Commit>): WorkItems {
    val workItems = mutableMapOf<WorkKey, MutableWorkKey>()

    for (commit in commits) {
        val workKeys = commit.workKeys.ifEmpty { listOf(WorkKey.Unknown) }
        for (workKey in workKeys) {
            workItems
                .getOrPut(workKey) { MutableWorkKey(workKey, commit.date) }
                .add(commit)
        }
    }

    return WorkItems(workItems.values.map(MutableWorkKey::toWorkItem))
}

private class MutableWorkKey(
    private val workKey: WorkKey,
    private var firstCommitDate: ZonedDateTime
) {
    private var linesAdded = 0
    private var linesRemoved = 0
    private var lastCommitDate: ZonedDateTime = firstCommitDate
    private val filesChanged = mutableSetOf<Path>()
    private val fileTouchCount = mutableMapOf<Path, Int>()
    private val absoluteChurnByAuthor = mutableMapOf<Author, Int>()
    private val absoluteChurnByType = mutableMapOf<CommitType, Int>()
    private var commitCount = 0
    private val collaborators = mutableSetOf<Author>()

    fun add(commit: Commit) {
        commitCount++
        collaborators.add(commit.author)
        collaborators.addAll(commit.coAuthors)

        var commitLinesChanged = 0
        for (fileChange in commit.fileChanges) {
            linesAdded += fileChange.additions
            linesRemoved += fileChange.deletions
            val path = Path.of(fileChange.path)
            filesChanged.add(path)
            fileTouchCount[path] = fileTouchCount.getOrDefault(path, 0) + 1
            commitLinesChanged += fileChange.additions + fileChange.deletions
        }

        absoluteChurnByAuthor[commit.author] = absoluteChurnByAuthor.getOrDefault(commit.author, 0) + commitLinesChanged
        absoluteChurnByType[commit.commitType] = absoluteChurnByType.getOrDefault(commit.commitType, 0) + commitLinesChanged

        val commitDate = commit.date
        if (commitDate.isBefore(firstCommitDate)) {
            firstCommitDate = commitDate
        }
        if (commitDate.isAfter(lastCommitDate)) {
            lastCommitDate = commitDate
        }
    }

    fun toWorkItem(): WorkItem = WorkItem(
        workKey = workKey,
        linesAdded = linesAdded,
        linesRemoved = linesRemoved,
        firstCommitDate = firstCommitDate,
        lastCommitDate = lastCommitDate,
        filesChanged = filesChanged.toSet(),
        contributions = absoluteChurnByAuthor
            .map { (author, lines) -> AuthorContribution(author, lines) }
            .sortedByDescending { it.linesChanged },
        absoluteChurnByType = absoluteChurnByType.toMap(),
        commits = commitCount,
        collaborators = collaborators.size,
        reworkFiles = fileTouchCount.filter { it.value > 1 }.keys
    )
}

File: visualization/src/mining/internal/work-item-miner.ts
import type {Author, Commit, CommitType, WorkKey} from '../api-types/commit.ts';
import {type AuthorContribution, type WorkItem, WorkItems} from "../api-types/work-items.ts";

interface MutableWorkItem {
  workKey: WorkKey;
  linesAdded: number;
  linesRemoved: number;
  firstCommitDate: Date;
  lastCommitDate: Date;
  filesChanged: Set<string>;
  fileTouchCount: Map<string, number>;
  absoluteChurnByAuthor: Map<string, number>;
  absoluteChurnByType: Map<CommitType, number>;
  commits: number;
  collaborators: Set<string>;
}

export class WorkItemMiner {
  private readonly workItems: Map<string, MutableWorkItem> = new Map();

  process(commit: Commit): void {
    let workKeys = commit.workKeys;
    if (workKeys.length === 0) {
      workKeys = [{ type: 'unknown' }];
    }

    for (const workKey of workKeys) {
      const key = this.workKeyToString(workKey);
      const item = this.getOrCreate(key, workKey, commit.date);
      this.addCommitToItem(item, commit);
    }
  }

  getResult(): WorkItems {
    const items = Array.from(this.workItems.values()).map(m => this.toWorkItem(m));
    return new WorkItems(items);
  }

  private getOrCreate(key: string, workKey: WorkKey, date: Date): MutableWorkItem {
    let item = this.workItems.get(key);
    if (!item) {
      item = {
        workKey,
        linesAdded: 0,
        linesRemoved: 0,
        firstCommitDate: date,
        lastCommitDate: date,
        filesChanged: new Set(),
        fileTouchCount: new Map(),
        absoluteChurnByAuthor: new Map(),
        absoluteChurnByType: new Map(),
        commits: 0,
        collaborators: new Set()
      };
      this.workItems.set(key, item);
    }
    return item;
  }

  private addCommitToItem(item: MutableWorkItem, commit: Commit): void {
    item.commits++;
    item.collaborators.add(this.authorKey(commit.author));
    for (const coAuthor of commit.coAuthors) {
      item.collaborators.add(this.authorKey(coAuthor));
    }

    let commitLinesChanged = 0;
    for (const fc of commit.fileChanges) {
      item.linesAdded += fc.additions;
      item.linesRemoved += fc.deletions;
      item.filesChanged.add(fc.path);
      item.fileTouchCount.set(fc.path, (item.fileTouchCount.get(fc.path) ?? 0) + 1);
      commitLinesChanged += fc.additions + fc.deletions;
    }

    const authorKey = this.authorKey(commit.author);
    item.absoluteChurnByAuthor.set(
      authorKey,
      (item.absoluteChurnByAuthor.get(authorKey) ?? 0) + commitLinesChanged
    );
    item.absoluteChurnByType.set(
      commit.commitType,
      (item.absoluteChurnByType.get(commit.commitType) ?? 0) + commitLinesChanged
    );

    if (commit.date < item.firstCommitDate) {
      item.firstCommitDate = commit.date;
    }
    if (commit.date > item.lastCommitDate) {
      item.lastCommitDate = commit.date;
    }
  }

  private toWorkItem(m: MutableWorkItem): WorkItem {
    const contributions: AuthorContribution[] = [];
    for (const [authorKey, lines] of m.absoluteChurnByAuthor) {
      const [name, email] = authorKey.split(':');
      contributions.push({ author: { name, email }, linesChanged: lines });
    }
    contributions.sort((a, b) => b.linesChanged - a.linesChanged);

    const reworkFiles: string[] = [];
    for (const [path, count] of m.fileTouchCount) {
      if (count > 1) {
        reworkFiles.push(path);
      }
    }

    return {
      workKey: m.workKey,
      linesAdded: m.linesAdded,
      linesRemoved: m.linesRemoved,
      firstCommitDate: m.firstCommitDate,
      lastCommitDate: m.lastCommitDate,
      filesChanged: Array.from(m.filesChanged),
      contributions,
      absoluteChurnByType: m.absoluteChurnByType,
      commits: m.commits,
      collaborators: m.collaborators.size,
      reworkFiles
    };
  }

  private workKeyToString(key: WorkKey): string {
    return key.type === 'known' ? key.key : '(unknown)';
  }

  private authorKey(author: Author): string {
    return `${author.name}:${author.email}`;
  }
}
